File: python/bien.py
from datetime import datetime, timedelta

class Bien:
    def __init__(self, nom: str, type_: str):
        self.nom = nom
        self.type = type_
        self.etat = "DISPONIBLE"
        self.date_fin_reservation = None

    def est_disponible(self) -> bool:
        self.mettre_a_jour_etat()
        return self.etat == "DISPONIBLE"

    def reserver(self, duree_jours: int):
        if duree_jours < 1:
            raise ValueError("La durée minimale de réservation est d’au moins 1 jour.")
        if not self.est_disponible():
            raise Exception(f"Le {self.type} '{self.nom}' est déjà réservé.")
        self.etat = "RESERVEE"
        self.date_fin_reservation = datetime.now() + timedelta(days=duree_jours)

    def annuler_reservation(self):
        if self.est_disponible():
            raise Exception(f"Le {self.type} '{self.nom}' n’est pas réservé.")
        self.etat = "DISPONIBLE"
        self.date_fin_reservation = None

    def mettre_a_jour_etat(self):
        if self.etat == "RESERVEE" and self.date_fin_reservation and datetime.now() > self.date_fin_reservation:
            self.etat = "DISPONIBLE"
            self.date_fin_reservation = None

    def __str__(self):
        self.mettre_a_jour_etat()
        date_str = f" jusqu’au {self.date_fin_reservation.date()}" if self.date_fin_reservation else ""
        return f"{self.type.capitalize()}{{nom='{self.nom}', etat={self.etat}{date_str}}}"

File: php/Bien.php
<?php

class Bien {
    private string $nom;
    private string $type;
    private string $etat;
    private ?DateTime $dateFinReservation;

    public function __construct(string $nom, string $type) {
        $this->nom = $nom;
        $this->type = $type;
        $this->etat = "DISPONIBLE";
        $this->dateFinReservation = null;
    }

    public function getNom(): string {
        return $this->nom;
    }

    public function estDisponible(): bool {
        $this->mettreAJourEtat();
        return $this->etat === "DISPONIBLE";
    }

    public function reserver(int $dureeJours) {
        if ($dureeJours < 1) {
            throw new Exception("La durée minimale de réservation est d'au moins 1 jour.");
        }
        if (!$this->estDisponible()) {
            throw new Exception("Le {$this->type} '{$this->nom}' est déjà réservé.");
        }
        $this->etat = "RESERVEE";
        $this->dateFinReservation = (new DateTime())->modify("+{$dureeJours} days");
    }

    public function annulerReservation() {
        if ($this->estDisponible()) {
            throw new Exception("Le {$this->type} '{$this->nom}' n'est pas réservé.");
        }
        $this->etat = "DISPONIBLE";
        $this->dateFinReservation = null;
    }

    private function mettreAJourEtat() {
        if ($this->etat === "RESERVEE" && $this->dateFinReservation && new DateTime() > $this->dateFinReservation) {
            $this->etat = "DISPONIBLE";
            $this->dateFinReservation = null;
        }
    }

    public function __toString(): string {
        $this->mettreAJourEtat();
        $dateStr = ($this->etat === "RESERVEE" && $this->dateFinReservation)
            ? " jusqu’au " . $this->dateFinReservation->format('Y-m-d')
            : "";
        return ucfirst($this->type) . "{nom='{$this->nom}', etat={$this->etat}{$dateStr}}";
    }
}
